Sources/Vehicle/Templates/CarPlayEntitiesListTemplate.swift

```diff
@@ -5,65 +5,112 @@ import PromiseKit
+    func templateWillDisappear(template: CPTemplate) {
+        if self.template == template {
+            entitiesSubscriptionToken?.cancel()
+        }
+    }
+
+    func templateWillAppear(template: CPTemplate) {
+        if self.template == template {
+            update()
             entitiesSubscriptionToken = entitiesCachedStates.subscribe { [weak self] _, _ in
-                self?.updateListItems()
+                self?.update()
             }
         }
-
-        return template
     }
 
-    private func updateListItems() {
-        guard let entities = entitiesCachedStates.value, let listTemplate = template as? CPListTemplate else { return }
+    func update() {
+        guard let entities = entitiesCachedStates.value else { return }
+
+        let entitiesFiltered = entities.all.filter { entity in
+            switch filterType {
+            case let .domain(domain):
+                return entity.domain == domain
+            case let .areaId(entityIdsAllowed):
+                if let domain = Domain(rawValue: entity.domain) {
+                    return entityIdsAllowed.contains(entity.entityId) && domain.isCarPlaySupported
+                } else {
+                    return false
+                }
+            }
+        }
 
-        let entitiesFiltered = entities.all.filter { $0.domain == domain }
-        let entitiesSorted = entitiesFiltered
-            .sorted(by: { $0.attributes.friendlyName ?? $0.entityId < $1.attributes.friendlyName ?? $1.entityId })
+        let entitiesSorted = entitiesFiltered.sorted(by: { e1, e2 in
+            let lowPriorityStates: Set<String> = [Domain.State.unknown.rawValue, Domain.State.unavailable.rawValue]
+            let state1 = e1.state
+            let state2 = e2.state
+
+            let deviceClassOrder: [HAEntity.DeviceClass] = [.garage, .gate]
+
+            if lowPriorityStates.contains(state1), !lowPriorityStates.contains(state2) {
+                return false
+            } else if lowPriorityStates.contains(state2), !lowPriorityStates.contains(state1) {
+                return true
+            } else {
+                if deviceClassOrder.contains(e1.deviceClass), !deviceClassOrder.contains(e2.deviceClass) {
+                    return true
+                } else if deviceClassOrder.contains(e2.deviceClass), !deviceClassOrder.contains(e1.deviceClass) {
+                    return false
+                }
+            }
 
-        let startIndex = currentPage * itemsPerPage
-        let endIndex = min(startIndex + itemsPerPage, entitiesSorted.count)
-        let entitiesToShow = Array(entitiesSorted[startIndex ..< endIndex])
+            return (e1.attributes.friendlyName ?? e1.entityId) < (e2.attributes.friendlyName ?? e2.entityId)
+        })
+
+        // Prevent unecessary update and UI glitch for non-touch screen CarPlay
+        let entitiesIds = entitiesSorted.map(\.entityId).sorted()
+        guard entitiesIdsCurrentlyInList != entitiesIds else {
```

Comment:
does this ever match when pagination is required?

i think it may be easier to match the entities here in a dictionary to a `CPListItem`, e.g.

```
// at a class level
var entityToListItem: [String: CPListItem]

// in this method
for entity in entities {
  let listItem: CPListItem
  if let existing = entityToListItem[entity.entityId] {
    listItem = existing
  } else {
    listItem = CPListItem()
    entityToListItem[entity.entityId] = listItem
  }
  // do updates / configure

}
```

i think you would need to move the "and change the items displayed" into the paginator to really invalidate/reset the template

Reply:
That is sort of what is happening already, the the list of entityIds change it means that some entity was added or removed, which means the paginated list needs an update, otherwise if the entityIds list is the same it means the state of an or multiple entities were updated, then those specific items are updated IF they are visible at that moment.

in CPListTemplate there is also not the concept of inserting a new item, so when we have a new entity we have to update the whole section

Reply:
I see the intent but I think pagination is making this "if" never work for cases where pagination happens at all.

Reply:
This check happens before everything is paginated, it's goal is just to prevent to recreate the list when the only thing that has changed was the state of entities. Let's say a new entity is added to your HA instance, then entitiesIds will have one more item and therefore different from entitiesIdsCurrentlyInList (previously set), which then will lead to the list being recreated.